InformationScripting/src/queries/Join.cpp

```diff
@@ -0,0 +1,155 @@
+** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**
+***********************************************************************************************************************/
+
+#include "Join.h"
+
+#include "../query_framework/QueryRegistry.h"
+
+namespace InformationScripting {
+
+const QStringList Join::VALUE_ARGUMENT_NAMES{"v", "values"};
+const QStringList Join::AS_ARGUMENT_NAMES{"a", "as"};
+const QStringList Join::ON_ARGUMENT_NAMES{"o", "on"};
+
+Optional<TupleSet> Join::executeLinear(TupleSet input)
+{
+	// First check which values we want to join on
+
+	const QRegularExpression onRegex{"(\\w+)\\.(\\w+)=(\\w+)\\.(\\w+)"};
+	auto onMatches = onRegex.match(arguments_.argument(ON_ARGUMENT_NAMES[1]));
+	if (!onMatches.hasMatch())
+		return {"Join requires the on argument in form tag.value=tag2.value"};
+
+	std::pair<std::pair<QString, QString>, std::pair<QString, QString>> joinOn{
+		{onMatches.captured(1), onMatches.captured(2)}, {onMatches.captured(3), onMatches.captured(4)}};
+
+	QList<std::pair<QString, QString>> values;
+	QString tag1;
+	QString tag2;
+	for (const auto& value : arguments_.argument(VALUE_ARGUMENT_NAMES[1]).split(",", QString::SkipEmptyParts))
+	{
+		auto parts = value.split(".", QString::SkipEmptyParts);
+		if (parts.size() <= 1)
+			return {"join values need to be fully specified with a dot"};
+		if (parts.size() > 2)
+			return {"join values can only have a single . (dot)"};
+		values.push_back({parts[0], parts[1]});
+
+		if (tag1.isNull() || tag1 == parts[0]) tag1 = parts[0];
+		else if (tag1 != parts[0] && tag2.isNull()) tag2 = parts[0];
+		else if (tag2 != parts[0]) return {"join values can only be from 2 different tuples"};
+	}
+	if (tag1.isNull() || tag2.isNull())
+		return {"join only works on 2 different tuples"};
+
+	auto tag1Tuples = input.tuples(tag1);
+	auto tag2Tuples = input.tuples(tag2);
+	if (tag1Tuples.isEmpty() || tag2Tuples.isEmpty())
+		return {input, "Not enough input for join"};
+
+	if (tag1 != joinOn.first.first)
+		std::swap(joinOn.first, joinOn.second);
+
+	if (tag1 != joinOn.first.first || tag2 != joinOn.second.first)
+		return {"join: Tags in values have to match with the tags in the on argument"};
+
+	auto id1Name = joinOn.first.second;
+	auto id2Name = joinOn.second.second;
+
+	for (const auto& tuple1 : tag1Tuples)
+	{
+		// the id always has to be a string, or some other statically defined value. :(
```

Comment:
This and the next line belong to line 89 just below. And in fact I just realized that it is plain wrong. I thought we had to extract the Property value to do the comparison, but we can just use Property::operator== so it will work for any property value.